`SortingAnalyzer` core documentation (#2563)
SpikeInterface/spikeinterface: doc/modules/core.rst

SortingAnalyzer
---------------

The :py:class:`~spikeinterface.core.SortingAnalyzer` class is the core object to combine a
:py:class:`~spikeinterface.core.BaseRecording` and a :py:class:`~spikeinterface.core.BaseSorting` object.
This is the first step for additional analyses, and the basis of several postprocessing and quality metrics
computations.

> The :py:class:`~spikeinterface.core.SortingAnalyzer` is provides a convenient API to access the underlying

Reviewer: ```suggestion
The :py:class:`~spikeinterface.core.SortingAnalyzer` provides a convenient API to access the underlying
```

:py:class:`~spikeinterface.core.BaseSorting` and :py:class:`~spikeinterface.core.BaseRecording` information,
and it supports several extensions (derived from the :py:class:`~spikeinterface.core.AnalyzerExtension` class)
to perform further analysis.

Importantly, the :py:class:`~spikeinterface.core.SortingAnalyzer` handles the *sparsity* (see [Sparsity]_ section), i.e.,
the channels on which waveforms and templates are defined on, for example, based on a physical distance from the
channel with the largest peak amplitude.

> Most of the extensions live in the :code:`postprocessing` module, but there are some *core* extensions too:

Reviewer: ```suggestion
Most of the extensions live in the :code:`postprocessing` module (with the :code:`quality_metrics` extension in the :code:`qualitymetrics` module) , but there are some *core* extensions too:
```

I think we need to be upfront here about the quality metrics since the design choice was made to separate qc from postprocessing we probably want to highlight multiple times that to get all functionality we need to import both modules. But I'm fine with my wording being changed if we want to rewrite this sentence.

> Most of the extensions live in the :code:`postprocessing` module, but there are some *core* extensions to:

Reviewer: ```suggestion
Most of the extensions live in the :code:`postprocessing` module, but there are some *core* extensions too:
```

> * select random spikes for downstream analysis (e.g., extracting waveforms or fit a PCA model)

Reviewer: ```suggestion
* select random spikes for downstream analysis (e.g., extracting waveforms or fitting a PCA model)
```

* estimate templates
* extract waveforms for single spikes
* compute channel-wise noise levels

.. note::

>     Some extensions depend on others, which must be pre-computed. For example:code:`waveforms`, the :code:`waveforms` extension depends on

Reviewer: ```suggestion
    Some extensions depend on others, which must be pre-computed. For example :code:`waveforms`, the :code:`waveforms` extension depends on
```

    the :code:`random_spikes` one. If the latter is not computed, computing :code:`waveforms` will throw an error.

.. code-block:: python

>     from spikeinterface import create_sorting_analyzer

Reviewer: ```suggestion

    from spikeinterface import create_sorting_analyzer
```

I think for sphynx you need one return right between the block and the code.

    # create in-memory sorting analyzer object
    sorting_analyzer = create_sorting_analyzer(
        sorting=sorting,
        recording=recording,
        sparse=True, # default
        format="memory", # default
    )

    print(sorting_analyzer)

.. code-block:: bash

    >>> SortingAnalyzer: 4 channels - 10 units - 1 segments - memory - sparse - has recording

>     >>> Loaded 0 extenstions:

Reviewer: We need to fix this typo in the source code. the repr/str should say extensions and not extenstions....

Reviewer: L192 in sortinganalyzer.py to be fixed. It won't let me put a comment there to mark it out.

Author: Fixed in last commit

The :code:`sorting_analyzer` object implements convenient functions to access the underlying :code:`recording` and

> :code:`sorting` objects information:

Reviewer: ```suggestion
:code:`sorting` objects' information:
```

.. code-block:: python

    num_channels = sorting_analyzer.get_num_channels()
    num_units = sorting_analyzer.get_num_units()

>     sampling_frequency = sorting_analyzer.sampling_frequency

Reviewer: ```suggestion
    sampling_frequency = sorting_analyzer.sampling_frequency()
```

We could do this as the function instead of the attribute right, just to make everything be a function for the docs?

Reviewer 2: we also have get_sampling_frequncy() but I think it is OK to show attributes also no ?

Reviewer: It's just a pedagogical difference. For a beginner they have to learn both the library and the difference between a function and an attribute. So my style is more beginner-friendly in that we remove them having to learn a bunch of python features by saying look everything has a function to grab a piece of data. Right now if I read this I would ask myself my sampling_frequency is special and doesn't have a function whereas all the others have functions and that might distract me from the rest of the docs.

Doing a mix is better for the intermediate learning because they already know enough basics that they focus on the breadth of techniques. So they see that they can both grab the attribute or use a function and can chose their style. 

So for me it would be do we want this documentation to be more beginner or more intermediate friendly. My inclination is that people that read the docs need a bit more beginner friendly. So we give them one good solution and then once they improve their own python skills the'll realize pretty quickly that they can use the attribute if they want.

But this is a super small point.

Reviewer 3: I am not convinced that functions are more beginner friendly than attributes.   Consistency I think is a good argument though.

Reviewer: My argument was completely for consistency. I don't care if one does functions or attributes just be consistent and since most of them were already functions I chose the route of making all functions, but absolutely would be fine to just grab them all as attributes too (from my perspective).

Author: I normally use the attribute, which I think is easier. Maybe we can show both?

Reviewer: Our argument (Heberto and mine since I think we fall in the same camp for documentation in general) is that if we include both we are making new learners/users learn both that Python has attributes + functions and the spikeinterface schema (why is it sampling_frequency and not samping_rate or why frequency and not period) so we are teaching them two things at once rather than teach them one thing (ie spikeinterface stuff). In our schema for docs you teach them one thing super well and then they can learn the other cool Python tricks themselves (function vs attribute). I think you and Sam like showing how many features the project has which is great for more experienced users/developers but I have a feeling that they don't read the docs (for example Aurélien once said he has never looked at the docs). 

My example would be as you learn English you could learn "need", "must", "require", which all have slightly different flavors so I could teach you that they all accomplish the same thing + teach you their different connotations or I could just teach you the word "need" and make sure you really understand that and even if there is a slightly better word "need" would still get the job done.

Author: Then in this case I would rather use `sorting_analyzer.sampling_frequency`, since this is the preferred use.

Reviewer: Again it's a super small point and not a big deal for this PR :) Just something to keep in mind. We had a big discussion of some of this stuff here https://github.com/SpikeInterface/spikeinterface/issues/2327

    total_duration = sorting_analyzer.get_total_duration()
    ### NOTE ###
    # 'segment_index' is required for multi-segment objects
    num_samples = sorting_analyzer.get_num_samples(segment_index=0)

    # channel_ids and unit_ids
    channel_ids = sorting_analyzer.channel_ids
    unit_ids = sorting_analyzer.unit_ids

Once the :code:`sorting_analyzer` is instantiated, additional extensions can be computed:

.. code-block:: python
    # compute some additional extensions

    # create in-memory sorting analyzer object
    random_spikes_extension = sorting_analyzer.compute("random_spikes")
    templates_extension = sorting_analyzer.compute("fast_templates")
    waveforms_extension = sorting_analyzer.compute("waveforms")

    # each extension has a .data field: a dictionary with computed data
    print(templates_extension.data.keys())

.. code-block:: bash

    >>> dict_keys(['average'])

# TODO: extension __repr__

.. code-block:: python
    # arguments can be passed directly to the compute function
    # note that re-computing an extension will overwrite the existing one
    waveform_extension_2 = sorting_analyzer.compute("waveforms", ms_before=2, ms_after=5)

>     # multiple extensions can be computedwithin the same `compute` call

Reviewer: ```suggestion
    # multiple extensions can be computed within the same `compute` call
```

>     # multiple extensions can be computed within the same `compute` call

Reviewer: Do we want a note that there isn't an optimizer and so if doing this one must put them in the correct order (ie random_spikes must go first)?

Reviewer 2: I think at this level I would avoid to explain this complexity no ?
Maybe we can open a section on depencies after no ?

Reviewer: I think that is a good idea. Separate section so that we don't get too complex right! I do think we need to mention it because it will be frustration and not clear why the function fails if they do the wrong order. So it is important but definitely separate section.

    sorting_analyzer.compute(
        ["random_spikes", "waveforms", "templates", "noise_levels"]
    )

The :py:class:`~spikeinterface.core.SortingAnalyzer` by default is defined *in memory*, but it can be saved at any time
(or upon instantiation) to one of the following backends:

* | :code:`zarr`: the sorting analyzer is saved to a [Zarr]() folder, and each extension is a Zarr group. This is the

>   | recommended backend, since Zarr files can be writter to/read from the cloud and compression is applied.

Reviewer: ```suggestion
  | recommended backend, since Zarr files can be written to/read from the cloud and compression is applied.
```

* | :code:`binary_folder`: the sorting analyzer is saved to a folder, and each extension creates a sub-folder. The extension

>   | data are saved to either :code:`npy` (for arrays), `csv` (for dataframes), or `pickle` (for everything else).

Reviewer: ```suggestion
  | data are saved to either :code:`npy` (for arrays), :code:`csv` (for dataframes), or :code:`pickle` (for everything else).
```

The :code:`SortingAnalyzer.save_as` function will save the object **and all its extension** to disk.

.. code-block:: python
    # create a "processed" folder
    processed_folder = Path("processed")

    sorting_analyzer_zarr = sorting_analyzer.save_as(
        folder=processed_folder / "sorting_analyzer.zarr",
        format="zarr"
    )
    sorting_analyzer_binary = sorting_analyzer.save_as(
        folder=processed_folder / "sorting_analyzer_bin",
        format="binary_folder"
    )
    # sorting_analyzer_zarr and sorting_analyzer_binary are valid SortingAnalyzers,
    # now associated to a Zarr storage / binary folder backend

    # create sorting analyzer associated to a backend upon instantiation

>     # create in-memory sorting analyzer object

Reviewer: Am I misreading this. Why is it backend and in-memory for this comment?

Reviewer 2: Yes I think it is a mistake

>     # create sorting analyzer associated to a backend upon instantiation
>     # create in-memory sorting analyzer object

Reviewer 2: ```suggestion
    # We can also create directly a  SortingAnalyzer associated to a backend upon instantiation
    # for instance, this create "zarr" SortingAnalyzer object
```

    sorting_analyzer_with_backend = create_sorting_analyzer(
        sorting=sorting,
        recording=recording,
        sparse=True,
        format="zarr",
        folder="my-sorting-analyzer.zarr"
    )

Once a :code:`SortingAnalyzer` object is saved to disk, it can be easily reloaded with:

.. code-block:: python

    sorting_analyzer = si.load_sorting_analyzer(folder="my-sorting-analyzer.zarr")

.. note::

    When saved to disk, the :code:`SortingAnalyzer` will store a copy of the :code:`Sorting`` object,

>     because that is relatively small and needed for most (if not all!) operations. The same is not

Reviewer: ```suggestion
    because it is relatively small and needed for most (if not all!) operations. The same is not
```

    true for the :code:`Recording` object, for which only the main properties will be stored (e.g,
    :code:`sampling_frequency`, :code:`channel_ids`, :code:`channel_locations`, etc.) and
    a provenance to reload the :code:`Recording`. When loading a :code:`SortingAnalyzer` from disk,

>     an attempt is made to re-instantiate the :code:`Recording` object from the provenance. In case

Reviewer: ```suggestion
    an attempt is made to re-instantiate the :code:`Recording` object from the provenance. In cases
```

Either in cases of failure...
or in the case of failure.

* estimate templates
* extract waveforms for single spikes
* compute channel-wise noise levels

.. note::

    Some extensions depend on others, which must be pre-computed. For example :code:`waveforms`, the :code:`waveforms` extension depends on
    the :code:`random_spikes` one. If the latter is not computed, computing :code:`waveforms` will throw an error.

.. code-block:: python

    from spikeinterface import create_sorting_analyzer

    # create in-memory sorting analyzer object
    sorting_analyzer = create_sorting_analyzer(
        sorting=sorting,
        recording=recording,
        sparse=True, # default
        format="memory", # default
    )

    print(sorting_analyzer)

.. code-block:: bash

    >>> SortingAnalyzer: 4 channels - 10 units - 1 segments - memory - sparse - has recording
    >>> Loaded 0 extensions:

The :code:`sorting_analyzer` object implements convenient functions to access the underlying :code:`recording` and
:code:`sorting` objects' information:

.. code-block:: python

    num_channels = sorting_analyzer.get_num_channels()
    num_units = sorting_analyzer.get_num_units()

>     sampling_frequency = sorting_analyzer.sampling_frequency

Author: ```suggestion
    sampling_frequency = sorting_analyzer.get_sampling_frequency()
    # or: sampling_frequency = sorting_analyzer.sampling_frequency
```

.. code-block:: bash

    >>> dict_keys(['average'])

# TODO: extension __repr__

.. code-block:: python
    # arguments can be passed directly to the compute function
    # note that re-computing an extension will overwrite the existing one
    waveform_extension_2 = sorting_analyzer.compute("waveforms", ms_before=2, ms_after=5)

    # multiple extensions can be computed within the same `compute` call
    sorting_analyzer.compute(
        ["random_spikes", "waveforms", "templates", "noise_levels"]
    )

The :py:class:`~spikeinterface.core.SortingAnalyzer` by default is defined *in memory*, but it can be saved at any time
(or upon instantiation) to one of the following backends:

* | :code:`zarr`: the sorting analyzer is saved to a [Zarr]() folder, and each extension is a Zarr group. This is the
  | recommended backend, since Zarr files can be written to/read from the cloud and compression is applied.
* | :code:`binary_folder`: the sorting analyzer is saved to a folder, and each extension creates a sub-folder. The extension
  | data are saved to either :code:`npy` (for arrays), :code:`csv` (for dataframes), or :code:`pickle` (for everything else).

> The :code:`SortingAnalyzer.save_as` function will save the object **and all its extension** to disk.

Reviewer 3: Missing and s for the plural on the bold format.

> The :code:`SortingAnalyzer.save_as` function will save the object **and all its extension** to disk.

Author: ```suggestion
The :code:`SortingAnalyzer.save_as` function will save the object **and all its extensions** to disk.
```
